src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -215,109 +115,133 @@ internal void OnSolutionBuild(object sender, UIContextChangedEventArgs e)
             }
 
             var asyncToken = _listener.BeginAsyncOperation("OnSolutionBuild");
-            _taskQueue.ScheduleTask(() =>
+            _taskQueue.ScheduleTask(async () =>
             {
-                // build is done, remove live error and report errors
-                IDictionary<ProjectId, IList<DiagnosticData>> liveProjectErrors;
-                IDictionary<DocumentId, IList<DiagnosticData>> liveDocumentErrors;
-                GetLiveProjectAndDocumentErrors(out liveProjectErrors, out liveDocumentErrors);
+                // nothing to do
+                if (_state == null)
+                {
+                    return;
+                }
 
-                using (var documentIds = SharedPools.Default<List<DocumentId>>().GetPooledObject())
-                using (var projectIds = SharedPools.Default<List<ProjectId>>().GetPooledObject())
+                // we are about to update live analyzer data using one from build.
+                // pause live analyzer
+                var service = _workspace.Services.GetService<IGlobalOperationNotificationService>();
+                using (var operation = service.Start("BuildDone"))
                 {
-                    lock (_gate)
-                    {
-                        documentIds.Object.AddRange(_documentToDiagnosticsMap.Keys);
-                        projectIds.Object.AddRange(_projectToDiagnosticsMap.Keys);
-                    }
+                    // we will have a race here since we can't track version of solution the out of proc build actually used.
+                    // result of the race will be us dropping some diagnostics from the build to the floor.
```

Comment:
well, no other way since again build doesn't give us that info (which action produced the diagnostic).